src/main/kotlin/com/msd/robot/domain/Inventory.kt

```diff
@@ -0,0 +1,80 @@
+package com.msd.robot.domain
+
+import com.msd.domain.ResourceType
+import java.util.*
+import javax.persistence.ElementCollection
+import javax.persistence.Entity
+import javax.persistence.Id
+
+@Entity
+class Inventory {
+    @Id
+    val id = UUID.randomUUID()
+    var storageLevel: Int = 0
+        internal set(value) {
+            if (value > 5) throw UpgradeException("Max Storage Level has been reached. Upgrade not possible.")
+            else if (value > storageLevel + 1)
+                throw UpgradeException(
+                    "Cannot skip upgrade levels. Tried to upgrade from level $storageLevel to level $value"
+                )
+            else if (value <= storageLevel)
+                throw UpgradeException("Cannot downgrade Robot. Tried to go from level $storageLevel to level $value")
+            field = value
+        }
+    var maxStorage = UpgradeValues.storageByLevel[storageLevel]
+        private set
+    var usedStorage = 0
+        private set
+
+    @ElementCollection
+    private val resourceMap = mutableMapOf(
+        ResourceType.COAL to 0,
+        ResourceType.IRON to 0,
+        ResourceType.GEM to 0,
+        ResourceType.GOLD to 0,
+        ResourceType.PLATIN to 0,
+    )
+
+    /**
+     * Adds a resource to this inventory. The inventory can hold all resources simultaneously, but the amount of
+     * resources held cannot exceed <code>maxStorage</code>.
+     *
+     * @param resource  the resource which will be added to the inventory
+     * @param amount    the amount that will be added
+     */
+    fun addResource(resource: ResourceType, amount: Int) {
+        val newUsedStorage = usedStorage + amount
+        if (newUsedStorage > maxStorage) {
+            resourceMap[resource] = resourceMap[resource]!! + amount - (newUsedStorage - maxStorage)
+            usedStorage = maxStorage
+            throw InventoryFullException("Added resources exceed maxStorage. Would be ${newUsedStorage}, max is $maxStorage")
+        } else {
+            resourceMap[resource] = resourceMap[resource]!! + amount
+            usedStorage += amount
+        }
+    }
+
+    /**
+     * Returns the stored amount of a given resource. The resource will still remain in the inventory
+     *
+     * @param resource  the resource of which the amount should be returned
+     * @return          the stored amount of the resource as an <code>Int</code>
+     */
+    fun getStorageUsageForResource(resource: ResourceType): Int {
+        return resourceMap[resource]!!
+    }
+
+    /**
+     * Takes a specified amount of resources from this inventory. The resources will be removed from the inventory.
+     *
+     * @param resource  the resource which should be taken
+     * @param amount    the amount which should be taken
+     * @return          the taken amount of the resource
+     */
+    fun takeResource(resource: ResourceType, amount: Int): Boolean {
```

Comment:
This method breaks the contract it defines in the comment above. It should return the amount of resources taken, instead of a boolean.

Reply:
No, it should return a boolean. I forgot to change the doc.